src/core/ucc_progress_queue_mt.c

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+
+#include "config.h"
+#include "ucc_progress_queue.h"
+#include "utils/ucc_malloc.h"
+#include "utils/ucc_log.h"
+#include "utils/ucc_spinlock.h"
+#include "utils/ucc_atomic.h"
+
+typedef struct ucc_pq_mt {
+    ucc_progress_queue_t super;
+    ucc_spinlock_t       locked_queue_lock;
+    ucc_coll_task_t***   tasks;
+    uint8_t              which_pool;
+    ucc_list_link_t      locked_queue;
+    uint32_t             tasks_countrs[2];
+} ucc_pq_mt_t; // TODO the struct isn't a queue because not maintaining order, maybe another name
+
+#define LINE_SIZE 8 // Could be changed, but in tests performed great
+#define NUM_POOLS 2 // We need extactly two pools
+
+static void ucc_pq_mt_enqueue(ucc_progress_queue_t *pq, ucc_coll_task_t *task) {
+    ucc_pq_mt_t *pq_mt = ucc_derived_of(pq, ucc_pq_mt_t);
+    int i;
+    int which_pool = task->was_progressed ^(pq_mt->which_pool & 1);
+    for (i = 0; i < LINE_SIZE; i++) {
+        if (ucc_atomic_bool_cswap64((uint64_t *) &(pq_mt->tasks[which_pool][i]), 0, (uint64_t) task)) {
+            ucc_atomic_add32(&pq_mt->tasks_countrs[which_pool], 1);
+            return;
+        }
+    }
+    ucc_spin_lock(&pq_mt->locked_queue_lock);
+    ucc_list_add_tail(&pq_mt->locked_queue, &task->list_elem);
+    ucc_spin_unlock(&pq_mt->locked_queue_lock);
+}
+
+static void ucc_pq_mt_dequeue(ucc_pq_mt_t *pq_mt, ucc_coll_task_t **popped_task_ptr, int is_first_call) {
+    int i, curr_which_pool, which_pool;
+    curr_which_pool = pq_mt->which_pool; // extract value in the beginning of the function
+    which_pool = curr_which_pool & 1; // turn from even/odd -> bool
+    ucc_coll_task_t *popped_task = NULL;
+    if (pq_mt->tasks_countrs[which_pool]) {
+        for (i = 0; i < LINE_SIZE; i++) {
+            popped_task = pq_mt->tasks[which_pool][i];
+            if (popped_task) {
+                if (ucc_atomic_bool_cswap64((uint64_t *) &(pq_mt->tasks[which_pool][i]), (uint64_t) popped_task, 0)) {
+                    ucc_atomic_sub32(&pq_mt->tasks_countrs[which_pool], 1);
+                    *popped_task_ptr = popped_task;
+                    popped_task->was_progressed = 1;
+                    return;
+                } else {
+                    i = -1;
+                    // TODO - break statement was here on XCCL, seems like a (perf) mistake - we want the pop to give at least 1 task if exist
```

Comment:
i don't agree. You only reach that branch in case you found a task to pop, and when trying to perform the atomic operation it failed. this failure for sure caused by another thread that was accessing the queue in the meanwhile. for almost all deques operations there is an insertion accompanied to it (unless the task progress was done). which means, there is a high chance that an insertion occurred in the meanwhile. if an insertion did occur, it for sure will be inserted in the blank space you just finish scanning. that is why you need to go back.

Reply:
and i prefer to leave comment, just for some time until we will be sure everything is ok. as i said, in XCCL it was checked repeatedly, and here not so much. in case of future bug, this might be a good point to check.

Reply:
Does not convince me at all. suppose there were 3 tasks in the queue. at positions 3,4,5 (0,1,2 - empty). you looped through, then could not take at pos 3 with swap. Means another thread got it. Just conitnue to pos 4. There is no evidence that anything has been inserted. The speculation above does not prove anything. However,  w/o this, the code is simpler: no else branch, no reset of the loop variable (which is badly looking practice), and no comment that referrs  to XCCL (which is by the way tested even less than UCC, especially MT mode).